docs/developer-feature-tutorial.md

```diff
@@ -99,74 +165,113 @@ Now, run zulip terminal and open web app from a different account. Start composi
+
+If the conditions are satisfied, we display `<sender> is typing...` if the `op` is `start` and display a help message if the `op` is `stop`.
 
 There are two parts to updating a widget in urwid:
-* Changing the widget
-* Updating the screen
 
-This line of code,
-```python
-self._w.footer.set_text([
-                    ' ',
-                    ('code', user['full_name']),
-                    ' is typing...'
-                ])
-```
-changes the footer text, and this
+* Changing the widget, which is done by View
+* Updating the screen, which is done by Controller
+
+These are done by the highlighted lines accordingly in `set_footer_text`
+
 ```python
-self.controller.update_screen()
+@asynch
+def set_footer_text(self, text_list: Optional[List[Any]]=None, duration: Optional[float]=None) -> None:
+    if text_list is None:
+        text = self.get_random_help()
+    else:
+        text = text_list
+>>> self._w.footer.set_text(text)
+>>> self.controller.update_screen()
+    if duration is not None:
+        assert duration > 0
+        time.sleep(duration)
+        self.set_footer_text()
 ```
-updates the screen to display the changes. This fully implements the typing feature.
 
-### Writing tests
+## Writing tests
+
+We use `pytest` to write all our tests. We have to write new tests, modify current ones or delete old ones whenever needed.
 
-Now, we update the tests for `register_initial_desired_events` by adding `typing`
+For this feature, we update the tests for `register_initial_desired_events` by adding `typing`
 to the event types.
-```diff
-# tests/core/test_core.py
 
-        event_types = [
-            'message',
-            'update_message',
-            'reaction',
-+           'typing',
-        ]
-        controller.client.register.assert_called_once_with(
-                                   event_types=event_types,
-                                   apply_markdown=True)
+```python
+# tests/core/test_core.py
+  event_types = [
+      'message',
+      'update_message',
+      'reaction',
+      'subscription',
+>>>   'typing',
+      'update_message_flags',
+      'update_display_settings',
+  ]
 ```
 
-`test_handle_typing_event` in `test_ui.py` implements testing for `handle_typing_event`. Please read it to understand how to write tests for a new function in zulip terminal.
+and we create a new function `test_handle_typing_event` in `test_model.py` which implements testing for `handle_typing_event`.
+
+Please read it to understand how to write tests for a new function in Zulip Terminal.
+
+How to run these tests? Easy:
+
+* To run all tests, just type `pytest` in your terminal.
+* To run linting type `./tools/lint-all` in your terminal.
+
+**Note**: If you happen to be using `pipenv`, use `pipenv run pytest` and `pipenv run ./tools/lint-all` instead.
+
+## Send Pull Request
```

Comment:
I like that you've included this conclusion of the contribution flow here, but I'd suggest this could be a "What next?" type section instead. We don't expect the user to contribute what they've explored reading this document - it's implemented already! - but we could say something along the lines of "Now that the code is clean & tested, normally this would be the point at which you'd create a pull request, where you might do <steps>...." and then flow into looking for issues, working on a real (unimplemented) feature and so on.

I don't feel that the git steps below help too much - the git guide is already referenced, and one often needs to run those commands in different ways in different orders, as you learn to use git and tidy up commits.